File: scripts/test-frame-contract.mjs
/* global console, process */
/*
 * The reader's half of `packages/sdk-python/frame_contract.json`: the UI may not
 * read a `detail` key nobody declares it will be sent.
 *
 * R-50 broke in the other direction — the runner emitted `detail.loop`, the
 * release number did not move, and the 1.4.1 on PyPI never emitted it, so
 * `indexLoops` skipped every repetition frame and the loop track was absent with
 * no error anywhere. The emitter's half is asserted in Python; the floor that
 * forces the publish is asserted in `test-sdk-version-gate.mjs`. What is left is
 * this direction: a `detail.somethingNew` added here reaches production as a
 * feature that silently does nothing, because an undeclared key is one the runner
 * has never promised to send.
 *
 * This is a source scan, deliberately: `ComputeLabFrame.detail` is
 * `Record<string, unknown>` on purpose — a runner reporting a key this build has
 * never heard of must still render it as a chip rather than vanish — so there is
 * no type for the compiler to check. A scan is what is left, and the guard
 * against a scan that quietly matches nothing is
 * `assertTheScanFoundTheKeyThisIssueWasAbout` at the end.
 *
 * Two things this cannot see, both stated so a green result is not read as more
 * than it is. A read is a bare key with no kind attached, so it is checked
 * against every key the contract declares for *any* kind rather than the kind it
 * will actually arrive on — deleting `loop` from `repetition.required` slips past
 * here, and that specific guarantee is held by name in `test_frame_contract.py`,
 * which asserts a `for` loop's frames carry it whatever the declaration says.
 * And a key reached some way none of the three patterns below match — through a
 * helper taking `detail` as a parameter, say — is not seen at all; the patterns
 * cover the forms these files actually use, and adding a fourth means teaching
 * this scan first.
 */
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

const contract = JSON.parse(readFileSync(resolve('packages/sdk-python/frame_contract.json'), 'utf8'));

/* Files that turn a frame into what the player sees. A `detail` read anywhere
 * else in the UI would be outside the Lab and outside this contract. */
const READERS = [
  'packages/ui/src/components/ComputeLabScreen.tsx',
  'packages/ui/src/components/computeLab/stage.tsx',
  'packages/ui/src/components/computeLab/stageModel.ts',
];

// ── The declaration itself ──────────────────────────────────────────────────
const declared = new Set();
for (const [kind, keys] of Object.entries(contract.kinds)) {
  assert.deepEqual(Object.keys(keys).sort(), ['optional', 'required'], `${kind} must declare both required and optional`);
  for (const bucket of ['required', 'optional']) {
    assert.ok(Array.isArray(keys[bucket]), `${kind}.${bucket} must be a list`);
    assert.equal(new Set(keys[bucket]).size, keys[bucket].length, `${kind}.${bucket} repeats a key`);
    for (const key of keys[bucket]) declared.add(key);
  }
  const both = keys.required.filter(key => keys.optional.includes(key));
  assert.deepEqual(both, [], `${kind} declares ${both} as both required and optional`);
}
assert.match(contract.sinceVersion, /^\d+(\.\d+)*$/, 'sinceVersion must be a dotted release number');

// ── Every named `detail` read in the Lab UI ─────────────────────────────────
/*
 * Two forms reach a frame's detail, and both have to be found:
 *   frame.detail?.loop        — read straight off the frame
 *   const detail = frame.detail;  … detail?.loop
 * The second is why aliases are collected per file rather than assumed: the name
 * `detail` is also bound to an array of rendered entries in ComputeLabScreen, and
 * counting that array's `.length` as a frame key would fail this test for no
 * reason. An alias is only a name assigned *from* `.detail`.
 */
const reads = new Map(); // key → the files that read it
const record = (key, file) => reads.set(key, [...(reads.get(key) || []), file]);

for (const file of READERS) {
  const source = readFileSync(resolve(file), 'utf8');

  for (const match of source.matchAll(/\.detail\s*(?:\?\.|\.)\s*([A-Za-z_$][\w$]*)/g)) record(match[1], file);
  for (const match of source.matchAll(/\.detail\s*(?:\?\.)?\[\s*['"]([^'"]+)['"]\s*\]/g)) record(match[1], file);
  for (const match of source.matchAll(/(?:const|let|var)\s*\{([^}]*)\}\s*(?::[^=\n]+)?=\s*[\w$?.]+\.detail\b/g))
    for (const name of match[1].split(',').map(part => part.split(':')[0].trim()).filter(Boolean)) record(name, file);

  const aliases = [...source.matchAll(/(?:const|let|var)\s+([A-Za-z_$][\w$]*)\s*(?::[^=\n]+)?=\s*[\w$?.]+\.detail\s*(?:;|$|\n)/gm)]
    .map(match => match[1]);
  for (const alias of new Set(aliases))
    for (const match of source.matchAll(new RegExp(`\\b${alias}\\s*(?:\\?\\.|\\.)\\s*([A-Za-z_$][\\w$]*)`, 'g')))
      record(match[1], file);
}

for (const [key, files] of [...reads].sort()) {
  assert.ok(
    declared.has(key),
    `${files.join(', ')} reads detail.${key}, which packages/sdk-python/frame_contract.json does not declare. ` +
      'Declare it there, move sinceVersion to the release that will emit it, and raise MIN_PYTHON_SDK_VERSION to ' +
      'that release — otherwise every player on an older runtime gets a screen that draws nothing and says nothing.',
  );
}

// ── The scan has to have worked ─────────────────────────────────────────────
function assertTheScanFoundTheKeyThisIssueWasAbout() {
  // A regex that matches nothing passes every assertion above. `loop` is the key
  // R-50 was filed about and the one the track cannot be drawn without, so its
  // absence from the scan means the scan broke, not that the UI stopped reading it.
  assert.ok(
    reads.has('loop'),
    'this scan found no detail.loop read in the Lab UI — either the track was removed, or the scan no longer matches ' +
      'how these files reach a frame. Fix the scan before trusting a green result from it.',
  );
}
assertTheScanFoundTheKeyThisIssueWasAbout();

console.log(
  `Frame contract passed (declared since ${contract.sinceVersion}; the Lab UI reads ${[...reads.keys()].sort().join(', ')})`,
);

File: packages/sdk-python/tests/test_frame_contract.py
"""The runner emits the frame shape it declares, and declares the release it emits it in.

R-50: `detail["loop"]` reached the runner and never reached a release number. The
1.4.1 on PyPI and the deployed UI both said "1.4.1" and disagreed about the frame,
so the version gate compared equal and admitted a runtime that emits no loop
identity — and the loop track, which is drawn only from that identity, was absent
for every player with no error anywhere to say so.

These tests are the emitter's half of `frame_contract.json`. They fail when the
runner's `detail` keys drift from the declaration, which is the edit whose
invisibility was the actual defect. The reader's half — that the server's floor
sits at or above the declared release, and that the UI reads no key nobody emits —
is `scripts/test-frame-contract.mjs`; the *published* artifact's half is
`scripts/test-python-sdk-artifact-lifecycle.py`.
"""

import json
from pathlib import Path

from netcrawl.compute_lab_runner import FRAME_KINDS, execute
from netcrawl.version import __version__

CONTRACT = json.loads((Path(__file__).resolve().parents[1] / "frame_contract.json").read_text(encoding="utf-8"))
KINDS = CONTRACT["kinds"]
FRAME = CONTRACT["frame"]

# Between them these reach every kind the runner can emit, and every optional key
# it can attach. `test_the_fixtures_reach_every_declared_kind` is what stops a
# conformance pass from being vacuous when one of them stops running.
FIXTURES = {
    "every_construct": """class ProblemSolver:
    def solution(self, a, b):
        def double(x):
            return x * 2
        nums = [a, b]
        total = 0
        for value in nums:
            if value > 0:
                total = total + double(value)
            else:
                total = total - 1
        index = 0
        while index < len(nums):
            index = index + 1
            if index > 5:
                break
        for left, right in [(a, b)]:
            total = total + left + right
        try:
            bad = nums[99]
        except IndexError:
            total = total + 1
        finally:
            total = total + 0
        return total
""",
    # An error nobody catches: every frame it leaves reports it.
    "uncaught": """class ProblemSolver:
    def solution(self, a, b):
        nums = [a, b]
        return nums[99]
""",
    # The one way `block_exit` carries `error`: a release that happened anyway.
    "finally_while_unwinding": """class ProblemSolver:
    def solution(self, a, b):
        nums = [a]
        try:
            bad = nums[99]
        finally:
            a = a
        return a
""",
    # An `assert` has a `test` and no body, so its decision carries no `taken`.
    "assert_only": """class ProblemSolver:
    def solution(self, a, b):
        assert a > 0
        return a
""",
    # A target that names nothing, so its binding carries no `bindings`.
    "subscript_target": """class ProblemSolver:
    def solution(self, a, b):
        nums = [a, b]
        nums[0] = 9
        return nums[0]
""",
    # `break` and `continue` have no fields at all, which is what a bare step is.
    # Both have to actually run: a `break` the loop never reaches is compiled and
    # instrumented, and emits nothing.
    "break_and_continue": """class ProblemSolver:
    def solution(self, a, b):
        total = 0
        for i in range(4):
            if i == 0:
                continue
            if i == 2:
                break
            total = total + i
        return total
""",
    # Nested loops: the identity is per live instance, not per source location.
    "nested": """class ProblemSolver:
    def solution(self, a, b):
        total = 0
        for i in range(3):
            for j in range(2):
                total = total + i + j
        return total
""",
}


def frames_of(source, params={"a": 2, "b": 3}, names=["a", "b"]):
    result = execute({"source": source, "params": params, "parameterNames": names, "limits": {"maxEvents": 2000}})
    assert result["status"] in {"trace_ready", "runtime"}, f"{result['status']}: {result.get('error')}"
    return result["frames"]


def every_frame():
    for name, source in FIXTURES.items():
        for frame in frames_of(source):
            yield name, frame


def test_the_contract_declares_every_kind_the_runner_can_emit():
    # A kind the runner can produce but the declaration omits would be exempt
    # from every check below, which is the same silence R-50 was.
    assert set(KINDS) == set(FRAME_KINDS)


def test_the_fixtures_reach_every_declared_kind():
    reached = {frame["kind"] for _, frame in every_frame()}
    assert reached == set(KINDS), f"unreached: {sorted(set(KINDS) - reached)}"


def test_the_runner_emits_no_detail_key_the_contract_does_not_declare():
    for fixture, frame in every_frame():
        declared = set(KINDS[frame["kind"]]["required"]) | set(KINDS[frame["kind"]]["optional"])
        emitted = set(frame.get("detail") or {})
        assert emitted <= declared, (
            f"{fixture}: a {frame['kind']} frame carries {sorted(emitted - declared)}, which frame_contract.json "
            "does not declare. Declare it and move sinceVersion to the release that will carry it — a reader on an "
            "older release cannot see this key, and a version compare cannot tell you so."
        )


def test_every_required_key_is_on_every_frame_of_its_kind():
    for fixture, frame in every_frame():
        emitted = set(frame.get("detail") or {})
        missing = set(KINDS[frame["kind"]]["required"]) - emitted
        assert not missing, (
            f"{fixture}: a {frame['kind']} frame is missing {sorted(missing)}. Either the runner stopped emitting a "
            "key the UI is entitled to, or the key was never unconditional and belongs in `optional`."
        )


def test_the_runner_emits_no_top_level_field_the_contract_does_not_declare():
    declared = set(FRAME["required"]) | set(FRAME["optional"])
    for fixture, frame in every_frame():
        undeclared = set(frame) - declared
        assert not undeclared, (
            f"{fixture}: a {frame['kind']} frame carries top-level {sorted(undeclared)}, which frame_contract.json "
            "does not declare. Declare it and move sinceVersion — a reader on an older release never receives it, "
            "and `detail` was only half this wire."
        )


def test_every_required_top_level_field_is_on_every_frame():
    for fixture, frame in every_frame():
        missing = set(FRAME["required"]) - set(frame)
        assert not missing, (
            f"{fixture}: a {frame['kind']} frame is missing top-level {sorted(missing)}. Every one of these is `?:` on "
            "`ComputeLabFrame`, so the UI reads it as undefined and draws nothing rather than failing — which is how "
            "`types` went missing for every player without a single error."
        )


def test_no_step_frame_carries_a_field_only_the_terminal_frame_may():
    # `error` is assembled by app.py for the closing error/limit frame. A step
    # frame carrying it would make the run look terminal to the server, which
    # refuses exactly that shape.
    for fixture, frame in every_frame():
        overlap = set(frame) & set(FRAME["terminalOnly"])
        assert not overlap, f"{fixture}: a {frame['kind']} step frame carries terminal-only {sorted(overlap)}"


def test_every_variable_carries_the_type_chip_the_boxes_draw():
    """R-50's second face, and the reason `detail` alone was not the contract.

    `1a92eed` added `frame["types"]` in the same commit as `detail["loop"]`, and
    the 1.4.1 on PyPI emits neither. `stage.tsx` reads `frame?.types?.[name]` for
    the chip under each variable box; absent, it renders nothing and no error —
    which is why the `int` labels are missing from R-50's "before" screenshot and
    present in the "after" one, a difference the first delivery read as identical.
    """
    frames = frames_of(FIXTURES["nested"])
    assert all("types" in frame for frame in frames)
    for frame in frames:
        # Parallel to `locals`: a box with no type chip is the failure being pinned.
        assert set(frame["types"]) == set(frame["locals"]), frame
        assert all(isinstance(name, str) for name in frame["types"].values()), frame
    holding = next(frame for frame in reversed(frames) if frame["locals"])
    assert holding["types"] == {name: "int" for name in holding["locals"]}, holding


def test_a_for_loop_carries_the_loop_identity_the_track_is_drawn_from():
    """R-50 itself, as the screen reported it: 「重複次數 2」 and no track.

    `detail["iteration"]` and `detail["loop"]` are set in the same unconditional
    block, so this is what tells the two apart from the outside — a runner that
    counts repetitions and cannot say *which loop* they belong to is the one that
    reached production.
    """
    repetitions = [frame for frame in frames_of(FIXTURES["nested"]) if frame["kind"] == "repetition"]
    assert repetitions
    for frame in repetitions:
        assert isinstance(frame["detail"]["loop"], int), frame
        assert isinstance(frame["detail"]["iteration"], int), frame
    # Per live instance, not per source location: the inner loop opens again on
    # every outer iteration, so three outer runs of one inner loop are four ids.
    assert len({frame["detail"]["loop"] for frame in repetitions}) == 4


def test_a_loop_block_exit_closes_the_track_it_opened():
    frames = frames_of(FIXTURES["nested"])
    opened = {frame["detail"]["loop"] for frame in frames if frame["kind"] == "repetition"}
    closed = {frame["detail"]["loop"] for frame in frames if frame["kind"] == "block_exit" and "loop" in (frame.get("detail") or {})}
    assert closed <= opened, "a block_exit named a loop that never repeated"
    assert closed, "no loop reported its exit, so no track can ever be drawn as finished"


def test_the_release_is_not_older_than_the_shape_it_emits():
    # sinceVersion is the release that first emits the declaration above. A tree
    # whose own version sits below it is claiming to be a release that predates
    # the frames it is producing — the exact inconsistency R-50 shipped.
    def parts(version):
        return [int(piece) for piece in version.split(".")]

    assert parts(__version__) >= parts(CONTRACT["sinceVersion"]), (
        f"netcrawl-sdk {__version__} emits the frame shape declared for {CONTRACT['sinceVersion']}"
    )
